From 3574e63aede491f01a409aee3c2a25c63d46419f Mon Sep 17 00:00:00 2001
From: Jacky Chou <jacky_chou@aspeedtech.com>
Date: Thu, 7 May 2026 17:45:06 +0800
Subject: [PATCH] arm: mach-aspeed: ast2705: detect MAC recovery boot

Read the IO strap register once in arch_misc_init() and set
boot_device=mac when recovery mode selects MAC. Keep the existing
storage boot-device and secure boot environment handling on the same
strap value.

Upstream-Status: Backport [SDK v11.03 migration]
Signed-off-by: Jacky Chou <jacky_chou@aspeedtech.com>
Change-Id: Ia4461d3a7df9f244a458f225ceaec185a2cc4dd7
Signed-off-by: rupeshs <rupeshs@ami.com>
---
 arch/arm/mach-aspeed/ast2705/platform.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-aspeed/ast2705/platform.c b/arch/arm/mach-aspeed/ast2705/platform.c
index abf70b5ede..64655a4ca4 100644
--- a/arch/arm/mach-aspeed/ast2705/platform.c
+++ b/arch/arm/mach-aspeed/ast2705/platform.c
@@ -39,8 +39,14 @@ enum env_location env_get_location(enum env_operation op, int prio)
 int arch_misc_init(void)
 {
 	if (IS_ENABLED(CONFIG_ARCH_MISC_INIT)) {
-		if ((readl(ASPEED_IO_HW_STRAP1) & SCU_IO_HWSTRAP_EMMC)) {
-			if ((readl(ASPEED_IO_HW_STRAP1) & SCU_IO_HWSTRAP_UFS))
+		u32 strap = readl(ASPEED_IO_HW_STRAP1);
+
+		if (strap & SCU_IO_HWSTRAP_RECOVERY) {
+			if (strap & SCU_IO_HWSTRAP_MAC) {
+				env_set("boot_device", "mac");
+			}
+		} else if (strap & SCU_IO_HWSTRAP_EMMC) {
+			if (strap & SCU_IO_HWSTRAP_UFS)
 				env_set("boot_device", "ufs");
 			else
 				env_set("boot_device", "mmc");
@@ -49,7 +55,7 @@ int arch_misc_init(void)
 			spi_bootarg_config();
 		}
 
-		if ((readl(ASPEED_IO_HW_STRAP1) & SCU_IO_HWSTRAP_SECBOOT))
+		if (strap & SCU_IO_HWSTRAP_SECBOOT)
 			env_set("verify", "yes");
 		else
 			env_set("verify", "no");
-- 
2.43.0

